satellite/overlay/combinedcache.go

```diff
@@ -0,0 +1,86 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package overlay
+
+import (
+	"context"
+	"sync"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+)
+
+type addressInfo struct {
+	address   string
+	lastIP    string
+	transport pb.NodeTransport
+}
+
+// CombinedCache is a simple caching mechanism for overlaycache updates. It
+// provdes methods to help reduce calls to UpdateAddress and UpdateTime, but can
+// be extended for other calls in the future.
+type CombinedCache struct {
+	DB
+	addressLock  sync.RWMutex
+	addressCache map[storj.NodeID]*addressInfo
```

Comment:
any reason not to add the lock directly into addressInfo?

Reply:
I personally feel the code is simpler with the lock existing in a separate abstraction. It's mainly handling the lazy-creation of the nodeID-based locks that would complicated the code here.

Reply:
Also, it’s possible addressInfo has the appropriate granularity. There may not yet be an addressInfo to hold the mutex, for example. It also makes updating the addressInfo harder as you can’t just overwrite the map key, possibly leading to subtle races.

Reply:
You should still be able to overwrite the map key.

Reply:
Maybe I misunderstand the suggestion and how the logic would work with the mutex on the addressInfo itself, then. Specifically, I don’t see how
1. You could avoid a lock around mutating the map any better
2. When you would want to acquire the mutex on the addressInfo

Reply:
I meant something like this https://github.com/storj/storj/pull/2716/files, since we are racing with multiple processes anyways.

Reply:
And if you want to have better ordering you could add a information version number.

Reply:
I think that’s closer to what the original PR did, and I objected (sorry, privately) because it makes it so that the cache can become out of sync with the database. While that’s apparently not a problem, the fact that it’s not a problem is not necessarily obvious. One must understand subtle whole system details to prove to themselves that it’s ok, and I can imagine future changes that lose that property. Additionally, every reader of the code in the future hits that problem. Indeed, I think it has been discussed multiple times even in this PR.

For those reasons, I strongly prefer if there is synchronization to prevent the cache from having different state from the database.

Reply:
@egonelbre yes your PR is basically exactly what I had in the first iteration of this. The reason I have changed it is because there is a risk that the cached data is out of sync with the underlying DB. As Jeff says, it is difficult to reason about precisely where this could cause problems (or not), and this is why I introduced the `KeyLock` concept to guarantee that the cache and the db are always in sync.

Reply:
Ok, let's keep keylock.